File: Java/examples/http_utils_example.java
import http_utils.mod;
import http_utils.mod.HttpResponse;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

/**
 * HTTP 工具模块使用示例
 *
 * 编译和运行:
 * javac -cp . http_utils/mod.java examples/http_utils_example.java
 * java -cp . examples.http_utils_example
 */
public class http_utils_example {

    public static void main(String[] args) {
        System.out.println("=== HTTP Utils Example ===\n");

        // 1. URL 编码解码示例
        demoUrlEncoding();

        // 2. 构建查询字符串示例
        demoQueryString();

        // 3. 模拟 HTTP 请求（实际网络请求需要可用网络）
        // 取消注释以下行以测试实际网络请求
        // demoHttpRequests();

        System.out.println("\n=== Example Completed ===");
    }

    /**
     * URL 编码解码演示
     */
    private static void demoUrlEncoding() {
        System.out.println("1. URL Encoding/Decoding Demo");
        System.out.println("------------------------------");

        String original = "hello world@example.com 中文测试";
        String encoded = mod.urlEncode(original);
        String decoded = mod.urlDecode(encoded);

        System.out.println("Original: " + original);
        System.out.println("Encoded:  " + encoded);
        System.out.println("Decoded:  " + decoded);
        System.out.println("Match: " + original.equals(decoded));
        System.out.println();
    }

    /**
     * 查询字符串构建演示
     */
    private static void demoQueryString() {
        System.out.println("2. Query String Building Demo");
        System.out.println("------------------------------");

        // 构建查询字符串
        Map<String, String> params = new HashMap<>();
        params.put("name", "张三");
        params.put("age", "25");
        params.put("city", "Beijing Shanghai");

        String queryString = mod.buildQueryString(params);
        System.out.println("Query String: " + queryString);

        // 构建完整 URL
        String baseUrl = "https://api.example.com/users";
        String fullUrl = mod.buildUrl(baseUrl, params);
        System.out.println("Full URL: " + fullUrl);

        // 向已有查询参数的 URL 添加参数
        String existingUrl = "https://api.example.com/search?q=java";
        Map<String, String> extraParams = new HashMap<>();
        extraParams.put("page", "1");
        extraParams.put("limit", "20");
        String appendedUrl = mod.buildUrl(existingUrl, extraParams);
        System.out.println("Appended URL: " + appendedUrl);
        System.out.println();
    }

    /**
     * HTTP 请求演示（需要网络连接）
     */
    private static void demoHttpRequests() {
        System.out.println("3. HTTP Requests Demo");
        System.out.println("----------------------");

        // GET 请求示例
        try {
            System.out.println("\n--- GET Request ---");
            HttpResponse getResp = mod.get("https://httpbin.org/get");
            System.out.println("Status: " + getResp.statusCode);
            System.out.println("Success: " + getResp.isSuccess());
            System.out.println("Body preview: " + getResp.body.substring(0, Math.min(200, getResp.body.length())) + "...");
        } catch (IOException e) {
            System.out.println("GET request failed: " + e.getMessage());
        }

        // 带请求头的 GET 请求
        try {
            System.out.println("\n--- GET with Headers ---");
            Map<String, String> headers = new HashMap<>();
            headers.put("Accept", "application/json");
            headers.put("X-Custom-Header", "MyValue");

            HttpResponse getResp = mod.get("https://httpbin.org/headers", headers);
            System.out.println("Status: " + getResp.statusCode);
            System.out.println("Body preview: " + getResp.body.substring(0, Math.min(200, getResp.body.length())) + "...");
        } catch (IOException e) {
            System.out.println("GET with headers failed: " + e.getMessage());
        }

        // POST JSON 请求
        try {
            System.out.println("\n--- POST JSON ---");
            String jsonBody = "{\"name\":\"张三\",\"age\":25,\"email\":\"zhangsan@example.com\"}";

            HttpResponse postResp = mod.postJson("https://httpbin.org/post", jsonBody);
            System.out.println("Status: " + postResp.statusCode);
            System.out.println("Success: " + postResp.isSuccess());
            System.out.println("Body preview: " + postResp.body.substring(0, Math.min(200, postResp.body.length())) + "...");
        } catch (IOException e) {
            System.out.println("POST JSON request failed: " + e.getMessage());
        }

        // POST 表单请求
        try {
            System.out.println("\n--- POST Form ---");
            Map<String, String> formData = new HashMap<>();
            formData.put("username", "admin");
            formData.put("password", "secret123");

            HttpResponse postResp = mod.postForm("https://httpbin.org/post", formData);
            System.out.println("Status: " + postResp.statusCode);
            System.out.println("Success: " + postResp.isSuccess());
        } catch (IOException e) {
            System.out.println("POST Form request failed: " + e.getMessage());
        }

        // PUT 请求
        try {
            System.out.println("\n--- PUT JSON ---");
            String jsonBody = "{\"id\":1,\"name\":\"Updated Name\"}";

            HttpResponse putResp = mod.putJson("https://httpbin.org/put", jsonBody);
            System.out.println("Status: " + putResp.statusCode);
            System.out.println("Success: " + putResp.isSuccess());
        } catch (IOException e) {
            System.out.println("PUT request failed: " + e.getMessage());
        }

        // DELETE 请求
        try {
            System.out.println("\n--- DELETE ---");
            HttpResponse delResp = mod.delete("https://httpbin.org/delete");
            System.out.println("Status: " + delResp.statusCode);
            System.out.println("Success: " + delResp.isSuccess());
        } catch (IOException e) {
            System.out.println("DELETE request failed: " + e.getMessage());
        }

        // 带自定义超时的请求
        try {
            System.out.println("\n--- Custom Timeout ---");
            HttpResponse resp = mod.get("https://httpbin.org/delay/1", null, 5000, 5000);
            System.out.println("Status: " + resp.statusCode);
            System.out.println("Success: " + resp.isSuccess());
        } catch (IOException e) {
            System.out.println("Custom timeout request failed: " + e.getMessage());
        }
    }
}

File: C/examples/http_utils_example.c
/**
 * @file http_utils_example.c
 * @brief HTTP Client Utilities Usage Examples
 * @author AllToolkit
 * @version 1.0.0
 *
 * This file demonstrates how to use the HTTP Utils library
 * for making HTTP requests and manipulating URLs.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../http_utils/mod.h"

/* Helper function to print response */
static void print_response(HttpResponse *response) {
    if (!response) {
        printf("Error: No response received\n");
        return;
    }
    
    printf("Status: %d %s\n", response->status_code, 
           response->status_text ? response->status_text : "Unknown");
    printf("Success: %s\n", response->success ? "true" : "false");
    printf("Response Time: %.3f seconds\n", response->response_time);
    printf("URL: %s\n", response->url ? response->url : "N/A");
    
    if (response->body) {
        printf("Body Length: %zu bytes\n", response->body_length);
        printf("Body (first 500 chars):\n%.500s%s\n\n", 
               response->body,
               response->body_length > 500 ? "..." : "");
    } else {
        printf("No body\n\n");
    }
}

/* Helper function to print URL components */
static void print_url_components(UrlComponents *comp) {
    if (!comp) {
        printf("Invalid URL\n");
        return;
    }
    
    printf("  Scheme: %s\n", comp->scheme ? comp->scheme : "(none)");
    printf("  Host: %s\n", comp->host ? comp->host : "(none)");
    printf("  Port: %d\n", comp->port);
    printf("  Path: %s\n", comp->path ? comp->path : "(none)");
    printf("  Query: %s\n", comp->query ? comp->query : "(none)");
    printf("  Fragment: %s\n", comp->fragment ? comp->fragment : "(none)");
    printf("  Userinfo: %s\n", comp->userinfo ? comp->userinfo : "(none)");
}

/* ============================================================================
 * Example 1: URL Parsing
 * ============================================================================ */
static void example_url_parsing(void) {
    printf("\n========================================\n");
    printf("Example 1: URL Parsing\n");
    printf("========================================\n\n");
    
    const char *urls[] = {
        "https://api.example.com/v1/users?page=1&limit=10",
        "http://localhost:8080/api/health",
        "https://user:pass@example.com:8443/path#section",
        NULL
    };
    
    for (int i = 0; urls[i]; i++) {
        printf("Parsing: %s\n", urls[i]);
        UrlComponents *comp = url_parse(urls[i]);
        print_url_components(comp);
        url_components_free(comp);
        printf("\n");
    }
}

/* ============================================================================
 * Example 2: URL Building
 * ============================================================================ */
static void example_url_building(void) {
    printf("\n========================================\n");
    printf("Example 2: URL Building\n");
    printf("========================================\n\n");
    
    UrlComponents comp = {0};
    comp.scheme = "https";
    comp.host = "api.example.com";
    comp.path = "/v1/search";
    comp.query = "q=hello+world&page=1";
    
    char *url = url_build(&comp);
    printf("Built URL: %s\n\n", url);
    free(url);
    
    /* Build with port */
    comp.port = 8443;
    url = url_build(&comp);
    printf("URL with port: %s\n\n", url);
    free(url);
}

/* ============================================================================
 * Example 3: URL Encoding/Decoding
 * ============================================================================ */
static void example_url_encoding(void) {
    printf("\n========================================\n");
    printf("Example 3: URL Encoding/Decoding\n");
    printf("========================================\n\n");
    
    const char *strings[] = {
        "hello world",
        "test@example.com",
        "special chars: &=?#",
        "Unicode: 你好世界",
        NULL
    };
    
    for (int i = 0; strings[i]; i++) {
        char *encoded = url_encode(strings[i]);
        char *decoded = url_decode(encoded);
        
        printf("Original: %s\n", strings[i]);
        printf("Encoded:  %s\n", encoded);
        printf("Decoded:  %s\n", decoded);
        printf("Match: %s\n\n", strcmp(strings[i], decoded) == 0 ? "YES" : "NO");
        
        free(encoded);
        free(decoded);
    }
}

/* ============================================================================
 * Example 4: Query String Building
 * ============================================================================ */
static void example_query_string(void) {
    printf("\n========================================\n");
    printf("Example 4: Query String Building\n");
    printf("========================================\n\n");
    
    const char *keys[] = {"name", "email", "age"};
    const char *values[] = {"John Doe", "john@example.com", "30"};
    
    char *query = url_build_query_string(keys, values, 3);
    printf("Query string: %s\n\n", query);
    free(query);
    
    /* Add params to URL */
    char *url = url_add_params("https://api.example.com/users", keys, values, 3);
    printf("Full URL: %s\n\n", url);
    free(url);
}

/* ============================================================================
 * Example 5: HTTP GET Request
 * ============================================================================ */
static void example_http_get(void) {
    printf("\n========================================\n");
    printf("Example 5: HTTP GET Request\n");
    printf("========================================\n\n");
    
    printf("Fetching data from httpbin.org...\n\n");
    
    HttpResponse *response = http_get("https://httpbin.org/get", NULL);
    print_response(response);
    http_response_free(response);
}

/* ============================================================================
 * Example 6: HTTP POST with JSON
 * ============================================================================ */
static void example_http_post_json(void) {
    printf("\n========================================\n");
    printf("Example 6: HTTP POST with JSON\n");
    printf("========================================\n\n");
    
    const char *json_data = "{\"name\":\"John\",\"age\":30,\"city\":\"New York\"}";
    
    printf("Sending JSON: %s\n\n", json_data);
    
    HttpResponse *response = http_post_json("https://httpbin.org/post", json_data, NULL);
    print_response(response);
    http_response_free(response);
}

/* ============================================================================
 * Example 7: HTTP POST with Form Data
 * ============================================================================ */
static void example_http_post_form(void) {
    printf("\n========================================\n");
    printf("Example 7: HTTP POST with Form Data\n");
    printf("========================================\n\n");
    
    const char *form_data = "username=admin&password=secret123";
    
    printf("Sending form data: %s\n\n", form_data);
    
    HttpResponse *response = http_post_form("https://httpbin.org/post", form_data, NULL);
    print_response(response);
    http_response_free(response);
}

/* ============================================================================
 * Example 8: Custom Headers
 * ============================================================================ */
static void example_custom_headers(void) {
    printf("\n========================================\n");
    printf("Example 8: Custom Headers\n");
    printf("========================================\n\n");
    
    HttpRequestOptions *options = http_options_new();
    options->headers = http_headers_new();
    http_headers_add(options->headers, "Authorization", "Bearer token123");
    http_headers_add(options->headers, "X-Custom-Header", "CustomValue");
    http_headers_add(options->headers, "Accept", "application/json");
    
    printf("Sending request with custom headers...\n\n");
    
    HttpResponse *response = http_get("https://httpbin.org/headers", options);
    print_response(response);
    
    http_response_free(response);
    http_options_free(options);
}

/* ============================================================================
 * Example 9: HTTP DELETE Request
 * ============================================================================ */
static void example_http_delete(void) {
    printf("\n========================================\n");
    printf("Example 9: HTTP DELETE Request\n");
    printf("========================================\n\n");
    
    printf("Sending DELETE request...\n\n");
    
    HttpResponse *response = http_delete("https://httpbin.org/delete", NULL);
    print_response(response);
    http_response_free(response);
}

/* ============================================================================
 * Example 10: URL Validation
 * ============================================================================ */
static void example_url_validation(void) {
    printf("\n========================================\n");
    printf("Example 10: URL Validation\n");
    printf("========================================\n\n");
    
    const char *urls[] = {
        "https://example.com",
        "http://localhost:8080/api",
        "not a url",
        "example.com",
        "https://user:pass@example.com/path?query=1#frag",
        NULL
    };
    
    for (int i = 0; urls[i]; i++) {
        bool valid = url_is_valid(urls[i]);
        printf("URL: %-40s -> %s\n", urls[i], valid ? "VALID" : "INVALID");
    }
    printf("\n");
}

/* ============================================================================
 * Example 11: Extract URL Components
 * ============================================================================ */
static void example_extract_components(void) {
    printf("\n========================================\n");
    printf("Example 11: Extract URL Components\n");
    printf("========================================\n\n");
    
    const char *url = "https://api.example.com/v1/users?page=1";
    
    char *domain = url_get_domain(url);
    char *path = url_get_path(url);
    
    printf("URL: %s\n", url);
    printf("Domain: %s\n", domain ? domain : "(null)");
    printf("Path: %s\n\n", path ? path : "(null)");
    
    free(domain);
    free(path);
}

/* ============================================================================
 * Main Function
 * ============================================================================ */
int main(void) {
    printf("\n");
    printf("╔══════════════════════════════════════════════════════════╗\n");
    printf("║     AllToolkit - HTTP Utils Example Program              ║\n");
    printf("║     Version: %s                                         ║\n", http_utils_version());
    printf("╚══════════════════════════════════════════════════════════╝\n");
    
    /* URL manipulation examples */
    example_url_parsing();
    example_url_building();
    example_url_encoding();
    example_query_string();
    example_url_validation();
    example_extract_components();
    
    /* HTTP request examples (require network) */
    printf("\n========================================\n");
    printf("HTTP Request Examples (Network Required)\n");
    printf("========================================\n");
    printf("Note: These examples require internet connectivity\n");
    printf("and use httpbin.org for testing.\n\n");
    
    example_http_get();
    example_http_post_json();
    example_http_post_form();
    example_custom_headers();
    example_http_delete();
    
    printf("\n========================================\n");
    printf("All examples completed!\n");
    printf("========================================\n\n");
    
    return 0;
}